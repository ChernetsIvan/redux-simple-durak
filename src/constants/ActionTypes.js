export const START_GAME_CLICKED = 'START_GAME_CLICKED';
export const BEGIN_GAME_CLICKED = 'BEGIN_GAME_CLICKED';
