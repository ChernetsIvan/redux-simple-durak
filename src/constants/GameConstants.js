export const cardRanks = [
  { text: "6", cardValue: 1 },
  { text: "7", cardValue: 2 },
  { text: "8", cardValue: 3 },
  { text: "9", cardValue: 4 },
  { text: "10", cardValue: 5 },
  { text: "В", cardValue: 6 },
  { text: "Д", cardValue: 7 },
  { text: "К", cardValue: 8 },
  { text: "Т", cardValue: 9 }
];

export const cardSuits = [
  { suit: "Ч" },
  { suit: "К" },
  { suit: "П" },
  { suit: "Б" }
];

export const statusMessages = {
  PlayerAttack: "Ваш ход!",
  PlayerDefence: "Защищайтесь!",
  PlayerDiscard: "Беру! Подбрасывать будете?",
  PlayerLose: "Вы проиграли! Сыграем ещё раз ? :)",
  PlayerWin: "Вы выиграли! Поздравляю :)",
  GameResultDraw: "Ничья!"
};
