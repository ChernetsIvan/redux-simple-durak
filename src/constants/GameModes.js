export const PlayerAttack = "PlayerAttack";
export const PlayerDefence = "PlayerDefence";
export const PlayerDiscard = "PlayerDiscard"; //AI забирает: игрок подбрасывает ему карты
export const AiAttack = "AiAttack";
export const AiDefence = "AiDefence";
export const AiDiscard = "AiDiscard"; //Игрок забирает: AI подбрасывает ему карты

export const PlayerWin = "PlayerWin";
export const AIWin = "AIWin";
export const TheDraw = "TheDraw"; //Ничья