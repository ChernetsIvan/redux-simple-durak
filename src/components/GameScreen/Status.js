import React from "react";
import PropTypes from "prop-types";
import { statusMessages } from "./../../constants/GameConstants";
import * as gameModes from "./../../constants/GameModes";

export const Status = ({ gameMode }) => {
  let outText = null;
  switch (gameMode) {
    case gameModes.PlayerAttack:
    case gameModes.AiDefence: {
      outText = statusMessages.PlayerAttack;
      break;
    }
    case gameModes.PlayerDefence:
    case gameModes.AiAttack: {
      outText = statusMessages.PlayerDefence;
      break;
    }
    case gameModes.PlayerDiscard: {
      outText = statusMessages.PlayerDiscard;
      break;
    }

    case gameModes.PlayerWin: {
      outText = statusMessages.PlayerWin;
      break;
    }
    case gameModes.AIWin: {
      outText = statusMessages.PlayerLose;
      break;
    }
    case gameModes.TheDraw: {
      outText = statusMessages.GameResultDraw;
      break;
    }

    case gameModes.NoCardsAI: {
      outText = statusMessages.NoCardsAI;
      break;
    }
    default: {
      outText = null;
    }
  }

  let dash = <span>&mdash;</span>;
  return (
    <div className="col-auto">
      <h3 className="mt-2">{dash} {outText}</h3>
    </div>
  );
};

Status.propTypes = {
  gameMode: PropTypes.string.isRequired
};
