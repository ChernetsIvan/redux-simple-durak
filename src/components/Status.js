import React from 'react';
import PropTypes from 'prop-types';
import {statusMessages} from './../constants/GameConstants';
import * as gameModes from './../constants/GameModes';

class Status extends React.Component{
    render(){
        let outText = null;

        if(this.props.fullDeck.length===0 && !this.props.firstStart){   
            if(this.props.computerCards.length===0){
                outText = statusMessages.PlayerLose;
            }
            if(this.props.playerCards.length===0){
                outText = statusMessages.PlayerWin;
            }
            if(this.props.computerCards.length===0 && 
                this.props.playerCards.length===0){
                outText = statusMessages.GameResultDraw;
            }         
        }

        if(outText === null) {
            switch(this.props.gameMode){
                case gameModes.PlayerAttack || gameModes.AiDefence: {
                    outText = statusMessages.PlayerAttack;
                    break;
                }
                case gameModes.PlayerDefence || gameModes.AiAttack: {
                    outText = statusMessages.PlayerDefence;
                    break;
                }
                case gameModes.PlayerDiscard: {
                    outText = statusMessages.PlayerDiscard;
                    break;
                }
                default:{
                    outText = null;
                }
            }
        }
        
        let dash = <span>&mdash;</span>;
        return (
            <div className="col-auto">
                <h3 className="mt-2">{dash} {outText}</h3>
            </div>
        );
    }
}

export default Status;

Status.propTypes = {
    gameMode:PropTypes.string,
    computerCards: PropTypes.array,
    playerCards: PropTypes.array,
    fullDeck: PropTypes.array,
    firstStart: PropTypes.bool
};

