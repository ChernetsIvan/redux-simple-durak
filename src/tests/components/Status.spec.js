import React from "react";
import renderer from "react-test-renderer";
import Status from "./../../components/GameScreen/Status";
import * as gameModes from "./../../constants/GameModes";

describe("Status tests", () => {
  const emptyFunc = () => {};

  test("Ai win", () => {
    const component = renderer.create(
      <Status gameMode={gameModes.AIWin} onGameOver={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Player win", () => {
    const component = renderer.create(
      <Status gameMode={gameModes.PlayerWin} onGameOver={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("The Draw", () => {
    const component = renderer.create(
      <Status gameMode={gameModes.TheDraw} onGameOver={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Player Attack", () => {
    const component = renderer.create(
      <Status gameMode={gameModes.PlayerAttack} onGameOver={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Player Defence", () => {
    const component = renderer.create(
      <Status gameMode={gameModes.PlayerDefence} onGameOver={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Player Defence", () => {
    const component = renderer.create(
      <Status gameMode={gameModes.PlayerDefence} onGameOver={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Player Discard", () => {
    const component = renderer.create(
      <Status gameMode={gameModes.PlayerDiscard} onGameOver={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Ai have not cards", () => {
    const component = renderer.create(
      <Status gameMode={gameModes.NoCardsAI} onGameOver={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("return null", () => {
    const component = renderer.create(
      <Status gameMode={""} onGameOver={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  })

});
