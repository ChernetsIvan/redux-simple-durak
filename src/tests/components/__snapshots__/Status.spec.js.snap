// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status tests Ai have not cards 1`] = `
<div
  className="col-auto"
>
  <h3
    className="mt-2"
  >
    <span>
      —
    </span>
     
    У меня закончились карты!
  </h3>
</div>
`;

exports[`Status tests Ai win 1`] = `
<div
  className="col-auto"
>
  <h3
    className="mt-2"
  >
    <span>
      —
    </span>
     
    Вы проиграли! Сыграем ещё раз ? :)
  </h3>
</div>
`;

exports[`Status tests Player Attack 1`] = `
<div
  className="col-auto"
>
  <h3
    className="mt-2"
  >
    <span>
      —
    </span>
     
    Ваш ход!
  </h3>
</div>
`;

exports[`Status tests Player Defence 1`] = `
<div
  className="col-auto"
>
  <h3
    className="mt-2"
  >
    <span>
      —
    </span>
     
    Защищайтесь!
  </h3>
</div>
`;

exports[`Status tests Player Defence 2`] = `
<div
  className="col-auto"
>
  <h3
    className="mt-2"
  >
    <span>
      —
    </span>
     
    Защищайтесь!
  </h3>
</div>
`;

exports[`Status tests Player Discard 1`] = `
<div
  className="col-auto"
>
  <h3
    className="mt-2"
  >
    <span>
      —
    </span>
     
    Беру! Подбрасывать будете?
  </h3>
</div>
`;

exports[`Status tests Player win 1`] = `
<div
  className="col-auto"
>
  <h3
    className="mt-2"
  >
    <span>
      —
    </span>
     
    Вы выиграли! Поздравляю :)
  </h3>
</div>
`;

exports[`Status tests The Draw 1`] = `
<div
  className="col-auto"
>
  <h3
    className="mt-2"
  >
    <span>
      —
    </span>
     
    Ничья!
  </h3>
</div>
`;

exports[`Status tests return null 1`] = `
<div
  className="col-auto"
>
  <h3
    className="mt-2"
  >
    <span>
      —
    </span>
     
  </h3>
</div>
`;
